### DISCUSSION TITLE
Implement Inception Prompting

### DISCUSSION THREAD

**Discussion by arbdwj** [microsoft/autogen#1331] [Q&A]
Inception prompting, introduced in [CAMEL](https://arxiv.org/pdf/2303.17760.pdf), is a method to mitigate role flipping and conversation deviation between the LM-agents

**Comment by victordibia** [comment]
@rokosbasilisk 

Can you give an example of inception prompting and how you are thinking it might be applied to agents to AutoGen agents? 

I have schemed through the inception prompt examples in the paper and these are mostly system prompts you can supply to any AutoGen Agent. If you observing situations where the agents flip roles, you can add some prompting in the agent system_message


```python

# create an AssistantAgent named "assistant"
assistant = autogen.AssistantAgent(
    name="coding assistant",
    llm_config= gpt4_llm_config,
    system_message="Never forget you are a <ASSISTANT_ROLE> and I am a <USER_ROLE>. This assigns the chosen role to the assistant agent and provides the agent with information about the user’s role. Never flip roles! Never the user! This prevents agents from flipping roles. In some cases, we have observed the assistant and the user switching roles, where the assistant suddenly takes control and instructs the user, and the user follows those instructions."
)

```

In general, the autogen agents framework provide flexibility to control agent behaviour by modifying the system messages. You can also review the default system messages for the [userproxyagent](https://github.com/microsoft/autogen/blob/957ec00edd3eb8b4ec09604a66e68116aee8f412/autogen/agentchat/user_proxy_agent.py#L28C16-L28C16) and [assistantagent](https://github.com/microsoft/autogen/blob/957ec00edd3eb8b4ec09604a66e68116aee8f412/autogen/agentchat/assistant_agent.py#L16)

![Image](https://github.com/microsoft/autogen/assets/1547007/fb438848-cbea-429c-8b2d-58b4aa424d5d)

**Comment by afourney** [comment]
I'd love a metric on role flipping. I've seen it. It happens. But I don't have a great understanding of how often.
(And if all agents are GTP-4 anyways... how much it matters)

**Comment by hkboujrida** [comment]
i'm using open hermes mistral for UserProxyAgent, and AssistantAgent with codellama:python, some time i don't know who is who, i'm struggling to have consistent behaviour, role flipping, gratification loop and so on.
